File: services/ml-service/app/services/text_embeddings.py
"""
Text embedding generation using Sentence Transformers
"""
import torch
import numpy as np
from typing import List
from app.core.models import get_text_model
from app.core.config import settings


def mean_pooling(model_output, attention_mask):
    """
    Mean pooling - take attention mask into account for correct averaging
    """
    token_embeddings = model_output[0]  # First element of model_output contains all token embeddings
    input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)


async def generate_text_embeddings(texts: List[str]) -> List[List[float]]:
    """
    Generate embeddings for a list of texts
    
    Args:
        texts: List of text strings
        
    Returns:
        List of embedding vectors (normalized)
    """
    if not texts:
        return []
    
    model, tokenizer = get_text_model()
    device = torch.device(settings.DEVICE)
    
    # Tokenize texts
    encoded_input = tokenizer(
        texts,
        padding=True,
        truncation=True,
        max_length=512,
        return_tensors='pt'
    ).to(device)
    
    # Generate embeddings
    with torch.no_grad():
        model_output = model(**encoded_input)
    
    # Mean pooling
    embeddings = mean_pooling(model_output, encoded_input['attention_mask'])
    
    # Normalize embeddings
    embeddings = torch.nn.functional.normalize(embeddings, p=2, dim=1)
    
    # Convert to list
    embeddings_list = embeddings.cpu().numpy().tolist()
    
    return embeddings_list


async def generate_text_embedding(text: str) -> List[float]:
    """
    Generate embedding for a single text
    
    Args:
        text: Text string
        
    Returns:
        Embedding vector (normalized)
    """
    embeddings = await generate_text_embeddings([text])
    return embeddings[0]

File: services/model-worker/src/services/text-embeddings.ts
import { getTextModel } from '../models/loader.js';
import { env } from '../config/env.js';

/**
 * Generate text embeddings using the loaded model
 * 
 * @param texts - Array of text strings to embed
 * @returns Array of embedding vectors
 */
export async function generateTextEmbeddings(texts: string[]): Promise<number[][]> {
  if (texts.length === 0) {
    return [];
  }

  // Validate batch size
  if (texts.length > env.MAX_BATCH_SIZE) {
    throw new Error(`Batch size ${texts.length} exceeds maximum ${env.MAX_BATCH_SIZE}`);
  }

  const model = getTextModel();

  try {
    // Generate embeddings for all texts
    const output = await model(texts, {
      pooling: 'mean', // Mean pooling for sentence embeddings
      normalize: true, // L2 normalization
    });

    // Convert to arrays
    const embeddings: number[][] = [];
    
    for (let i = 0; i < texts.length; i++) {
      const embedding = Array.from(output[i]!.data as Float32Array);
      embeddings.push(embedding);
    }

    return embeddings;
  } catch (error) {
    console.error('Text embedding generation failed:', error);
    throw new Error(`Failed to generate text embeddings: ${error instanceof Error ? error.message : 'Unknown error'}`);
  }
}

/**
 * Generate single text embedding
 * 
 * @param text - Text string to embed
 * @returns Embedding vector
 */
export async function generateTextEmbedding(text: string): Promise<number[]> {
  const embeddings = await generateTextEmbeddings([text]);
  return embeddings[0]!;
}

